### DISCUSSION TITLE
Better GitHub Code Search for R with tree-sitter-r

### DISCUSSION THREAD

**Discussion by DavisVaughan** [community/community#120397] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hi GitHub Code Search team!

I'm an employee of Posit, best known for RStudio and our collection of R packages like dplyr, tidyr, and the rest of the tidyverse.

We'd really love to see improved code search for R, like how in these two C examples:
- You can click on a function call and it will tell you the definition site within the project
- You can search for a function name and it will bring up the definition site within the project

<img width="1078" alt="Screenshot 2024-04-20 at 8 09 54 AM" src="https://github.com/community/community/assets/19150088/e51dc5a3-5948-4c08-b770-e22488c8e1a1">
<img width="474" alt="Screenshot 2024-04-20 at 8 07 53 AM" src="https://github.com/community/community/assets/19150088/27d7e069-6f0f-4aae-9b5b-660db125285f">

We've written a tree-sitter grammar that we are pretty happy with. It's current at the `next` branch but that will be moving to `main` soon. https://github.com/r-lib/tree-sitter-r/tree/next

I imagine I probably need to write a `tags.scm` for you guys? i.e. https://tree-sitter.github.io/tree-sitter/code-navigation-systems. We do have `highlights.scm` already.

Is there anything else you need? Once we have `tags.scm`, what next?

And in general I'm curious about the process of how things are kept up to date. Like, if we make an update or improvement in tree-sitter-r, how could that flow into your Code Search? Would we manually ping you about it to get things synced up?

I've seen https://github.com/orgs/community/discussions/10934 by @dcreager and @colinwm, are there any other community guides that have been developed since then?

Thanks in advance!

**Comment by look** [comment, marked as answer]
@DavisVaughan thanks for the suggestion. We'd be happy to add code navigation support for R. You are correct that having a `tags.scm` will make this much easier for us. Once that is done and the code is in your mainline, please let us know. The next step for us is to take your `tags.scm` and adapt it to our system (our pipeline requires certain names for the language elements returned by the query). Then the next time we rebuild our code search index, R symbols will be included.

**Reply by DavisVaughan** [reply]
@look tree-sitter-r is ready for you to use! Everything is on `main` now, and pretty stable.

- Repo: https://github.com/r-lib/tree-sitter-r
- `tags.scm` https://github.com/r-lib/tree-sitter-r/blob/main/queries/tags.scm
- `highlights.scm` https://github.com/r-lib/tree-sitter-r/blob/main/queries/highlights.scm (if you use them)

I'm giving a talk in just under 1 month and was hoping to talk about this there, is it possible that a rebuild would happen before then? That would be so exciting for the R community!

Thanks so much!

**Reply by look** [reply]
Hi @DavisVaughan, I'm looking into what it would take to get your new R parser integrated. Can you please push a new release of the Rust crate to crates.io? The current version hasn't been updated in 2 years: https://crates.io/crates/tree-sitter-r

**Reply by DavisVaughan** [reply]
Sure I'll do one tomorrow thank you! I'll report back when done.

**Reply by DavisVaughan** [reply]
@look you should be good to go to get tree-sitter-r 1.0.1 from crates.io now! https://crates.io/crates/tree-sitter-r/versions

**Reply by DavisVaughan** [reply]
@look just wanted to check in and see if there was any hope at all of getting this in before July 6 (that's roughly when I give my conference talk)!

(I realize y'all have your own timelines, just need to plan accordingly.)

**Reply by look** [reply]
I'll do my best! Upgrading parsers can be challenging because they depend on different versions of `tree-sitter` but I'm working on it.

We recently open sourced some [documentation on how our code navigation system works](https://github.com/github/code-navigation). Of note to you is that we support fully-qualified names. I'm not an R expert, but I noticed it supports defining functions within functions. I'm not sure if there's a way to refer to these nested functions, but if there is, a [scoped query](https://github.com/github/code-navigation?tab=readme-ov-file#fully-qualified-names) will make the navigation look better. Let me know if that's something you'd want to add.

**Reply by DavisVaughan** [reply]
> open sourced some documentation on how our code navigation system works

Oh cool! I'll definitely give that a close read. The list of supported tags was definitely something I was just guessing on - but I'm happy to see I guessed right with `@definition.function` and `@reference.call` 😄.

> I noticed it supports defining functions within functions

You can certainly do something like this

```r
fn <- function(x) {
  helper <- function(x) {
    print(x)
  }

  for (elt in x) {
    helper(elt)
  }
}
```

But it doesn't seem like that is exactly what you are referring to in https://github.com/github/code-navigation?tab=readme-ov-file#fully-qualified-names.

IIUC, the closest thing we have in R that is widely used is probably an R6 class, with a class name and some public/private methods for that class. For example: https://github.com/r-lib/testthat/blob/f060ccec365e1343ae926c53c215a423b295d7a8/R/stack.R#L7. It would be used like:

```r
stack <- Stack$new()
stack$push(1)
```

But I'd argue that that is a very very small percentage of R code, so I don't think I would spend any time there right now. It is also not native to the language - it is a convention provided by an extension package, so it _might_ be out of scope for tree-sitter-r itself (we've tried to draw the line at supporting R syntax rather than community conventions so far).

**Reply by look** [reply]
Sounds good. That was my take as well from my limited R experience (and asking ChatGPT 😆). As it stands, your tagging query will extract nested functions. This example:

```R
fn <- function(x) {
  helper <- function(x) {
  }
}
```

Will result in code navigation for both `fn` and `helper`. If that's OK with you, I'll be able to use your tagging queries as-is.

**Reply by DavisVaughan** [reply]
That is what I understand as well, both `fn` and `helper` get tagged with plain old `definition.function`.

That is perfectly fine with us right now!

---

Happy to answer any other questions you might have!

**Reply by look** [reply]
@DavisVaughan I've deployed R symbol extraction. It will probably take a few weeks for the full rollout (which requires rebuilding all our search indexes), but the symbol navigator is now available in R files [like this one](https://github.com/r-lib/testthat/blob/main/R/compare.R). Thanks for your help with this!

**Reply by DavisVaughan** [reply]
That is extremely exciting!! Thank you so much for your help and timely responses.

I do have a few questions about it if you have a moment.

It does look like it is recognizing function definitions in some cases now. i.e. if I click on `vec_order`'s definition here, it finds it and now reports the `Definition` location, which is cool! I happened to take a before/after screenshot:

Before:


<img width="1522" alt="Screenshot 2024-06-26 at 8 50 48 AM" src="https://github.com/community/community/assets/19150088/0577f844-30c5-4612-be31-cd43b3916bd6">

After:

<img width="1496" alt="Screenshot 2024-06-26 at 8 51 12 AM" src="https://github.com/community/community/assets/19150088/573ab7f1-af52-4022-92ec-bb5d3be24a74">

---

But if I click on a _reference_ to `vec_order()`, nothing happens yet. Is that part of what still needs to be reindexed?

https://github.com/community/community/assets/19150088/f35800dc-8e75-4696-9313-05c80c09c368

I'm also expecting that if I use the search bar, then `vec_order()` will come up as an R `function` that I can jump to the definition of, which also doesn't happen yet (it just shows the C equivalent). Is that another thing that we are waiting on reindexing for?

https://github.com/community/community/assets/19150088/5ec7f915-08dd-4f88-bd57-c4ae45b2982b

Thanks so much!

**Reply by look** [reply]
Yes, you've got it. Those features are powered by search so they will start working after we reindex with the extraction enabled. The per file extraction doesn't use the index which is why it started working immediately.

**Reply by DavisVaughan** [reply]
@look this is a long shot but - there's no way to force a reindex of just a single repo is there? I was hoping to show https://github.com/rstudio/vetiver-r as an example of code search working now, but nothing has come online yet. Just trying to decide exactly what to say for the talk. If not, that's okay, I understand it takes a few weeks like you said. Thanks!

**Reply by look** [reply]
@DavisVaughan sorry if this is too late for your talk (I was OOO) but I have reindexed rstudio/vetiver-r for you.

**Reply by DavisVaughan** [reply]
Not too late! It looks like top level Search now includes R functions!! That is one of the major things I was going to show, so thank you so much!

The other thing I was potentially going to show (but I can leave it out if not possible to get it up in time) is _clicking_ on an R function reference, and then using the Symbols pane (which pops up for other languages) to jump straight to definition, but that still isn't working in vetiver-r. Any ideas? (i.e. when I click on `vetiver_write_plumber()` in the video below, was hoping the symbols pane would pop up). Either way, thank you so much for the extra time you're putting in here!

<img width="985" alt="Screenshot 2024-07-08 at 10 00 44 PM" src="https://github.com/community/community/assets/19150088/8c240c06-36f1-42e0-ac22-223d0981632e">

https://github.com/community/community/assets/19150088/4199a6d0-82dd-4284-9356-d139281a8a32

**Reply by DavisVaughan** [reply]
FYI my talk is actually this Thursday

**Reply by look** [reply]
@DavisVaughan I looked into this a little. I think there are a couple of problems. In the file you demo in your video, clicking works for _some_ items but not others (for example, it works pretty well for function arguments). This is because the calls don't have the right CSS class to be considered a symbol. I think we need to upgrade our syntax highlighter to fix that.

I also found that functions with `.` in the name don't work well (for example, `vetiver_create_description.default` in [`prepare.R`](https://github.com/rstudio/vetiver-r/blob/main/R/prepare.R)). This is because the syntax highlighter treats the whole text as a symbol, and the regex that we use to determine a symbol excludes symbols with `.` in them. I'm not sure what the correct behavior in this case is.

We will work on improving this but unfortunately it won't be ready for your talk tomorrow.

**Reply by DavisVaughan** [reply]
> clicking works for some items but not others (for example, it works pretty well for function arguments). This is because the calls don't have the right CSS class to be considered a symbol. I think we need to upgrade our syntax highlighter to fix that.

Let me know if there is anything I can do to help!

---

> the regex that we use to determine a symbol excludes symbols with . in them. I'm not sure what the correct behavior in this case is.

In R, in general you can have a `.` in the name of a variable, like

```r
> x.y <- 1
> x.y
[1] 1
```

In the case of

```r
vetiver_create_description.default <- function(model) {
```

this is a special case of using the `.` where this is a _default_ S3 method for the generic function `vetiver_create_description()`. Similarly you could define `vetiver_create_description.foo <- function(model) {` as the S3 method for `vetiver_create_description()` specific to `model` objects of class `"foo"`.

That's only considered an S3 method because `vetiver_create_description()` itself is defined as a generic function. In general you can even use a `.` in a function name, like:

```r
data.frame <- function(..., stringsAsFactors = FALSE) {
```

which is a function called `data.frame()`, not an S3 method for `data()` on objects of class `"frame"`.

This is obviously a confusing part of R, and I think it is fairly hard to figure out statically, so I think what I would do would be to unconditionally allow `.` as part of a symbol name, so `data.frame` and `vetiver_create_description.default` would be considered distinct symbols.

---

> unfortunately it won't be ready for your talk tomorrow

That is ok! I have the videos I need for the talk.

But just FYI, the day after you told me that search for vetiver-r was working, it...stopped working 😆 . Luckily I snagged a video before it stopped working, but here it is today (when it worked it would show a function called `vetiver_model` that you can jump to), not sure what's up with that:

https://github.com/community/community/assets/19150088/8dad39ea-83f0-44ad-94bf-d8a501b555c4

**Reply by DavisVaughan** [reply]
@look searching for functions using the search box has been working beautifully and looks to be rolled out across all R projects on GitHub now, thanks so much for your help!

My talk went very well, and people really really loved hearing about this. It even got a round of applause in the middle of the talk!

Is there any update regarding this?

> This is because the calls don't have the right CSS class to be considered a symbol. I think we need to upgrade our syntax highlighter to fix that.

Would love to be able to click on a reference to a function call and jump straight to the definition in the project
